### DISCUSSION TITLE
CTMRG with iterative (Lanczos) solver for SVD

### DISCUSSION THREAD

**Discussion by Yue-Zhengyuan** [QuantumKitHub/PEPSKit.jl#98] [Q&A] [answered]
Let $\chi$ be the dimension of the CTMRG environment. In the projector finding step of CTMRG, we know that only $\chi$ singular values (and their corresponding left/right singular vectors) are needed. Therefore, instead of doing a full SVD and then truncate, it can greatly improve the speed by using an iterative (Lanczos) algorithm for the SVD step, as noted in the paper ([arXiv:2308.12358](https://arxiv.org/abs/2308.12358), particularly when $\chi$ is large; see Fig. 16 there)

> Naumann, Jan, et al. "An introduction to infinite projected entangled-pair state methods for variational ground state simulations using automatic differentiation." SciPost Physics Lecture Notes (2024): 086.

It is stated that the algorithm is already in KrylovKit.jl. Can someone demonstrate how to use such an algorithm with CTMRG? I think it just amounts to setting `svd_alg = IterSVD()` in `CTMRG`, but it seems that `ProjectorAlg` has not been implemented for `IterSVD`. Thanks!

**Comment by pbrehmer** [comment, marked as answer]
Hey Yue, totally missed that question, sorry! I hope you figured this out in the meantime but in case this is still unclear, here is an example:

```julia
using KrylovKit
using TensorKit
using PEPSKit

Dbond = 2
χenv = 16
svd_alg = SVDAdjoint(; fwd_alg=IterSVD(; alg=GKL(; tol=1e-10)))
ctmrg_alg = SimultaneousCTMRG(; svd_alg)
ψ = InfinitePEPS(2, Dbond)
env = leading_boundary(CTMRGEnv(ψ, ℂ^χenv), ψ, ctmrg_alg);
```

I'll put an example in the docs once we do the documentation overhaul.

**Comment by Yue-Zhengyuan** [comment]
@pbrehmer Thanks! However, with symmetry or for fermions, I suppose it requires us to specify in advance the dimensions of each sector, and it cannot dynamically change them (similar to `FixedSpaceTruncation` for usual SVD)?

**Reply by pbrehmer** [reply]
Yes indeed. Not all truncation schemes are implemented or would even make sense straightaway in the context of `IterSVD`. Currently, only `NoTruncation` and `TruncationSpace` are implemented (so using `FixedSpaceTruncation` works just fine).

**Reply by pbrehmer** [reply]
If you're interested, you can find an example usage in `test/utility/svd_wrapper.jl`!
